nsg-netbox.py

```diff
@@ -51,96 +101,331 @@ def run(self):
+            version = res.headers.get("API-Version", "")
+        try:
+            return float(".".join(version.split("-")[0].split(".")[:2]))
+        except ValueError:
+            return None
 
-    def netbox_dcim(self, nbox):
-        if not self.config:
-            return nbox.dcim.devices.filter(status=1)
-        if not self.config.get('filters'):
-            self.config['filters'] = {'status': 1}
+    def get_netbox_devices(self):
+        result = {}
         filters = {}
-        for fk, fv in self.config['filters'].items():
-            if 'custom_fields' in fk:
-                fk = 'cf_{}'.format(fk.split('.')[1])
-            filters[fk] = fv
-        if self.config.get('whitelist'):
+        if self.config and self.config.get('filters'):
+            for fk, fv in self.config['filters'].items():
+                if 'custom_fields' in fk:
+                    fk = 'cf_{}'.format(fk.split('.')[1])
+                filters[fk] = fv
+        else:
+            filters = {"status": self.active}
+
+        if self.config and self.config.get('whitelist'):
             filters['tag'] = self.config['whitelist']
-        return nbox.dcim.devices.filter(**filters)
 
-    def get_primary_ip(self, device):
-        return str(ipaddress.ip_interface(device.primary_ip).ip)
+        devices = self.netbox_dcim(model="devices", filters=filters)
+        for device in devices:
+            ip = self.get_primary_ip(device)
+            if ip:
+                result[ip] = device
+            else:
+                host = f"{device.name}.{self.domain}" if self.domain and self.domain not in device.name else device.name
+                try:
+                    ip = socket.gethostbyname(host)
+                    if ip:
+                        result[ip] = device
+                    else:
+                        self.log.error("get_netbox_devices: can't resolve device name: {}".format(device.name))
+                except Exception as e:
+                    self.log.error("get_netbox_devices: can't resolve device name: {}: {}".format(device.name, e))
+            site = self._sites.get(device.site.name)
+            if isinstance(site, pynetbox.core.response.Record):
+                for key in ('facility', 'continent', 'metro', 'cluster'):
+                    if not hasattr(device, key) or not getattr(device, key):
+                        setattr(device, key, getattr(site, key) if hasattr(site, key) else None)
+            device.nsg_tags = self.extract_tags(device)
+        return result
+        # return {self.get_primary_ip(d): d for d in self.netbox_dcim(nbox) if self.condition(d)}
 
-    def condition(self, device):
-        if self.config.get('blacklist'):
+    def get_netbox_sites(self):
+        sites = self.netbox_dcim(model="sites")
+        for site in sites:
+            metro = site.name[:3]
+            try:
+                if not hasattr(site, "facility"):
+                    setattr(site, "facility", None)
```

Comment:
why do we do this manipulation of tags? This program is open source and is not being developed specifically for Dropbox. We should not make any assumption about tags that exist or not exist in Netbox, it is none of our business. This means there is no need to check if facility, cluster, etc, is there and try to back-fill them or do other checks.